### ISSUE TITLE
Cannot create azurerm_storage_share when firewall rules are enabled

### ISSUE THREAD

**Issue by grant-dennison** [terraform-providers/terraform-provider-azurerm#12411]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

```
PS C:\...> terraform -v
Terraform v1.0.1
on windows_amd64
+ provider registry.terraform.io/hashicorp/azurerm v2.65.0
+ provider registry.terraform.io/hashicorp/http v2.1.0
```

### Affected Resource(s)

* `azurerm_storage_account`
* `azurerm_storage_share`

### Terraform Configuration Files

```hcl
resource "azurerm_storage_account" "file_share_acc" {
  name                     = "failcreatesharest"
  resource_group_name      = azurerm_resource_group.rg.name
  location                 = azurerm_resource_group.rg.location
  account_tier             = "Premium"
  # Locally-redundant storage.
  account_replication_type = "LRS"
  account_kind             = "FileStorage"

  network_rules {
    bypass = ["AzureServices"]
    default_action = "Deny"
    ip_rules = [module.ip.ip]
    virtual_network_subnet_ids = [azurerm_subnet.subnet.id]
  }
}

# This is the resource that can't be created.
resource "azurerm_storage_share" "share" {
  name                 = "fail-create-share-sts"
  storage_account_name = azurerm_storage_account.file_share_acc.name
}

# The rest of this Terraform script is just setup for the above two resources
# so it is in the next collapsed block.
```

<details>
<summary>...the rest of the Terraform script</summary>

```hcl
terraform {
  backend "local" {}

  required_providers {
    azurerm = {
      source  = "hashicorp/azurerm"
      version = ">= 2.49.0"
    }
  }

  required_version = ">= 0.14.5"
}

provider "azurerm" {
  features {}
}

resource "azurerm_resource_group" "rg" {
  name     = "FAIL_CREATE_SHARE_RG"
  location = "East US 2"
}

resource "azurerm_virtual_network" "vnet" {
  name                = "fail_create_share_vnet"
  location            = "East US 2"
  resource_group_name = azurerm_resource_group.rg.name
  address_space       = ["10.0.0.0/16"]
}
resource "azurerm_subnet" "subnet" {
  name                 = "fail_create_share_subnet"
  resource_group_name  = azurerm_resource_group.rg.name
  virtual_network_name = azurerm_virtual_network.vnet.name
  address_prefixes     = ["10.0.1.0/24"]

  service_endpoints = [ "Microsoft.Storage" ]
}

# This module isn't necessary except to demonstrate that the IP we whitelist is our IP.
module "ip" {
  source  = "JamesWoolfenden/ip/http"
  version = "0.3.10"
}
```
</details>

<details>
<summary>.terraform.lock.hcl</summary>

```hcl
# This file is maintained automatically by "terraform init".
# Manual edits may be lost in future updates.

provider "registry.terraform.io/hashicorp/azurerm" {
  version     = "2.65.0"
  constraints = "2.65.0"
  hashes = [
    "h1:3jSw0J5yIjQfdwphFtOZylM1viYfWlVPp5E2URHb2HQ=",
    "zh:03ffb2151a9d8d6f3e7a5a1985cfecdb30b91781577b83be8fa75d050f75b2d8",
    "zh:3631760913bc9e4727de34cc5d54eb92661ced73b75fc50c21dd2af2aa0c71f2",
    "zh:36bee282d96334ecf8cd468dbd49ef56afbd9634db1b84f5df54f926ee17ce88",
    "zh:53e986e3458bf38013194147e44f067979f99d480436a2b845a3a1190fa9d05f",
    "zh:560595d18566666d1ae2cc9da0e0a321d549c9caa716f29ee670025ce6c5cb2a",
    "zh:8d6709d19211325ebc78f474741623c4c7c28407d109f3740ab008379a8e0f40",
    "zh:bf7f949faca42744048e85f6fa4fa72eb4815ef2e0bcca60dba7f5afbbc8f589",
    "zh:cd09a82e23656c9a1b0314b2480993f2cce21325ce2a595f434614f0ac5adc6b",
    "zh:d1b60c0d729e1a9e676e4962beaad5fc0f4c12e748cf1d8dcd2f4df6273dd3d1",
    "zh:e477221421a97a0f5e8111508b6d57469bd6d5cf2cc374ed08d31f7bbbdaaf20",
    "zh:f00f71f4a2514bcb49135b66a7bb3c0bb52c032a438277dc56606be36724b8bd",
  ]
}

provider "registry.terraform.io/hashicorp/http" {
  version = "2.1.0"
  hashes = [
    "h1:SE5ufHNXfkUaY6ZLLW742Pasb/Jx/Y/lUwoYS3XbElA=",
    "zh:03d82dc0887d755b8406697b1d27506bc9f86f93b3e9b4d26e0679d96b802826",
    "zh:0704d02926393ddc0cfad0b87c3d51eafeeae5f9e27cc71e193c141079244a22",
    "zh:095ea350ea94973e043dad2394f10bca4a4bf41be775ba59d19961d39141d150",
    "zh:0b71ac44e87d6964ace82979fc3cbb09eb876ed8f954449481bcaa969ba29cb7",
    "zh:0e255a170db598bd1142c396cefc59712ad6d4e1b0e08a840356a371e7b73bc4",
    "zh:67c8091cfad226218c472c04881edf236db8f2dc149dc5ada878a1cd3c1de171",
    "zh:75df05e25d14b5101d4bc6624ac4a01bb17af0263c9e8a740e739f8938b86ee3",
    "zh:b4e36b2c4f33fdc44bf55fa1c9bb6864b5b77822f444bd56f0be7e9476674d0e",
    "zh:b9b36b01d2ec4771838743517bc5f24ea27976634987c6d5529ac4223e44365d",
    "zh:ca264a916e42e221fddb98d640148b12e42116046454b39ede99a77fc52f59f4",
    "zh:fe373b2fb2cc94777a91ecd7ac5372e699748c455f44f6ea27e494de9e5e6f92",
  ]
}
```

</details>

### Debug Output

`$env:TF_LOG="DEBUG"`
<details>
<summary>`terraform apply` output (final lines)</summary>

```
2021-06-29T13:12:08.458-0400 [DEBUG] provider.terraform-provider-azurerm_v2.65.0_x5.exe: Cache Miss - looking up the account key for storage account "failcreatesharest"..: timestamp=2021-06-29T13:12:08.455-0400
2021-06-29T13:12:08.458-0400 [DEBUG] provider.terraform-provider-azurerm_v2.65.0_x5.exe: AzureRM Request:
POST /subscriptions/REDACTED/resourceGroups/FAIL_CREATE_SHARE_RG/providers/Microsoft.Storage/storageAccounts/failcreatesharest/listKeys?%24expand=kerb&api-version=2021-01-01 HTTP/1.1
Host: management.azure.com
User-Agent: Go/go1.16.3 (amd64-windows) go-autorest/v14.2.1 Azure-SDK-For-Go/v55.3.0 storage/2021-01-01 HashiCorp Terraform/1.0.1 (+https://www.terraform.io) Terraform Plugin SDK/2.6.1 terraform-provider-azurerm/2.65.0 pid-REDACTED
Content-Length: 0
X-Ms-Correlation-Request-Id: REDACTED
Accept-Encoding: gzip: timestamp=2021-06-29T13:12:08.455-0400
2021-06-29T13:12:08.659-0400 [DEBUG] provider.terraform-provider-azurerm_v2.65.0_x5.exe: AzureRM Response for https://management.azure.com/subscriptions/REDACTED/resourceGroups/FAIL_CREATE_SHARE_RG/providers/Microsoft.Storage/storageAccounts/failcreatesharest/listKeys?%24expand=kerb&api-version=2021-01-01:
HTTP/2.0 200 OK
Cache-Control: no-cache
Content-Type: application/json
Date: Tue, 29 Jun 2021 17:12:08 GMT
Expires: -1
Pragma: no-cache
Server: Microsoft-Azure-Storage-Resource-Provider/1.0,Microsoft-HTTPAPI/2.0 Microsoft-HTTPAPI/2.0
Strict-Transport-Security: max-age=31536000; includeSubDomains
Vary: Accept-Encoding
X-Content-Type-Options: nosniff
X-Ms-Correlation-Request-Id: REDACTED
X-Ms-Ratelimit-Remaining-Subscription-Resource-Requests: REDACTED
X-Ms-Request-Id: REDACTED
X-Ms-Routing-Request-Id: CENTRALUS:20210629T171208Z:REDACTED

{"keys":[{"keyName":"key1","value":"REDACTED","permissions":"FULL"},{"keyName":"key2","value":"REDACTED","permissions":"FULL"}]}: timestamp=2021-06-29T13:12:08.658-0400
╷
│ Error: checking for existence of existing Storage Share "fail-create-share-sts" (Account "failcreatesharest" / Resource Group "FAIL_CREATE_SHARE_RG"): shares.Client#GetProperties: Failure responding to request: StatusCode=403 -- Original Error: autorest/azure: Service returned an error. Status=403 Code="AuthorizationFailure" Message="This request is not authorized to perform this operation.\nRequestId:REDACTED\nTime:2021-06-29T17:12:08.7126613Z"
│
│   with azurerm_storage_share.share,
│   on main.tf line 19, in resource "azurerm_storage_share" "share":
│   19: resource "azurerm_storage_share" "share" {
│
╵
2021-06-29T13:12:08.727-0400 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = transport is closing"
2021-06-29T13:12:08.751-0400 [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.terraform.io/hashicorp/azurerm/2.65.0/windows_amd64/terraform-provider-azurerm_v2.65.0_x5.exe pid=29336
2021-06-29T13:12:08.752-0400 [DEBUG] provider: plugin exited
```
</details>

### Expected Behaviour

The file share resource should be created in the storage account.

### Actual Behaviour

403 error:

```
azurerm_resource_group.rg: Creating...
azurerm_resource_group.rg: Creation complete after 1s [id=/subscriptions/REDACTED/resourceGroups/FAIL_CREATE_SHARE_RG]
azurerm_virtual_network.vnet: Creating...
azurerm_virtual_network.vnet: Creation complete after 5s [id=/subscriptions/REDACTED/resourceGroups/FAIL_CREATE_SHARE_RG/providers/Microsoft.Network/virtualNetworks/fail_create_share_vnet]
azurerm_subnet.subnet: Creating...
azurerm_subnet.subnet: Creation complete after 4s [id=/subscriptions/REDACTED/resourceGroups/FAIL_CREATE_SHARE_RG/providers/Microsoft.Network/virtualNetworks/fail_create_share_vnet/subnets/fail_create_share_subnet]
azurerm_storage_account.file_share_acc: Creating...
azurerm_storage_account.file_share_acc: Still creating... [10s elapsed]
azurerm_storage_account.file_share_acc: Still creating... [20s elapsed]
azurerm_storage_account.file_share_acc: Creation complete after 22s [id=/subscriptions/REDACTED/resourceGroups/FAIL_CREATE_SHARE_RG/providers/Microsoft.Storage/storageAccounts/failcreatesharest]
azurerm_storage_share.share: Creating...
╷
│ Error: checking for existence of existing Storage Share "fail-create-share-sts" (Account "failcreatesharest" / Resource Group "FAIL_CREATE_SHARE_RG"): shares.Client#GetProperties: Failure responding to request: StatusCode=403 -- Original Error: autorest/azure: Service returned an error. Status=403 Code="AuthorizationFailure" Message="This request is not authorized to perform this operation.\nRequestId:REDACTED\nTime:2021-06-29T16:54:27.3953379Z"
│
│   with azurerm_storage_share.share,
│   on main.tf line 19, in resource "azurerm_storage_share" "share":
│   19: resource "azurerm_storage_share" "share" {
│
```

### Steps to Reproduce

1. `terraform apply`
2. Confirm (with `yes`) resource creation

### Important Factoids

* Share can be created if `default_action` is changed to `"Allow"` (turning off the firewall).

### References

* #2977 is a similar issue, but it is for storage containers, not file shares.
* #8922 appears to be the same issue, but it was dismissed as a question rather than a bug.
* [#24677 against Terraform](https://github.com/hashicorp/terraform/issues/24677) is a similar issue with a partial workaround based on dependency ordering, but it isn't a perfect solution.
* Comment from @tombuildsstuff in #1764 suggests that including the runner IP in `ip_rules` should be sufficient, but my test case here demonstrates that this is not the case.
* There are some questions and answers on SO (such as [this](https://stackoverflow.com/a/64420330)) related to this issue, some of which include some partial workarounds.

**Comment by grant-dennison** [comment]
It may be the case that my runner machine is being routed to Azure differently (from internal IP) than to other services on the internet. This would mean that my whitelist IP entry doesn't match the request the blob storage account receives. Trying to work through a potentially related issue to get some clarity on this theory.

**Comment by grant-dennison** [comment]
I resolved my issue by just changing the region/location I used in the Terraform script. It seems that the VM running the Terraform script was in the same region I was trying to create these resources, so Azure was doing internal routing (using my local network IP) rather than public internet routing. It seems to me like there is still an issue to be resolved with that on Azure's end, but I don't think it's an issue with the Terraform provider.
